bot/cogs/tags.py

```diff
@@ -0,0 +1,265 @@
+
+        embed = Embed()
+        tags = []
+
+        tag_data = await self.get_tag_data(tag_name)
+
+        # If we found something, prepare that data
+        if tag_data:
+            embed.colour = Colour.blurple()
+
+            if tag_name:
+                embed.title = tag_name
+                self.tag_cooldowns[tag_name] = {
+                    "time": time.time(),
+                    "channel": ctx.channel.id
+                }
+
+            else:
+                embed.title = "**Current tags**"
+
+            if isinstance(tag_data, list):
+                tags = [f"**»**   {tag['tag_name']}" for tag in tag_data]
+                tags = sorted(tags)
+
+            else:
+                embed.description = tag_data['tag_content']
+
+        # If not, prepare an error message.
+        else:
+            embed.colour = Colour.red()
+            embed.title = "There are no tags in the database!"
+
+            if isinstance(tag_data, dict):
+                embed.description = f"Unknown tag: **{tag_name}**"
+
+            if tag_name:
+                embed.set_footer(text="To show a list of all tags, use bot.tags.get()")
+                embed.title = "Tag not found!"
+
+        # Paginate if this is a list of all tags
+        if tags:
+            return await paginate(
+                (lines for lines in tags),
+                ctx, embed,
+                footer_text="To show a tag, type bot.tags.get <tagname>",
+                empty=False,
+                max_size=200
+            )
+
+        return await ctx.send(embed=embed)
+
+    @with_role(ADMIN_ROLE, OWNER_ROLE, MODERATOR_ROLE)
+    @command(name="tags.set()", aliases=["tags.set", "tags.add", "tags.add()", "tags.edit", "tags.edit()", "add_tag"])
+    async def set_command(self, ctx: Context, tag_name: str, tag_content: str):
+        """
+        Create a new tag or edit an existing one.
+
+        :param ctx: discord message context
+        :param tag_name: The name of the tag to create or edit.
+        :param tag_content: The content of the tag.
+        """
+
+        embed = Embed()
+        embed.colour = Colour.red()
+
+        if "\n" in tag_name:
+            embed.title = "Please don't do that"
+            embed.description = "Don't be ridiculous. Newlines are obviously not allowed in the tag name."
+
+        else:
+            if not (tag_name and tag_content):
+                embed.title = "Missing parameters!"
+                embed.description = "The tag needs both a name and some content"
+                return await ctx.send(embed=embed)
+
+            tag_name = tag_name.lower()
+            tag_data = await self.post_tag_data(tag_name, tag_content)
+
+            if tag_data.get("success"):
+                embed.colour = Colour.blurple()
+                embed.title = "Tag successfully added!"
+                embed.description = f"**{tag_name}** added to tag database."
+            else:
+                embed.title = "Database error",
```

Comment:
Some of the embed titles have punctuation but some don't. It might be more consistent to add periods to these?